### ISSUE TITLE
Use provider to interpret X509 critical extension?

### ISSUE THREAD

**Issue by taavi-valjaots** [openssl/openssl#16290]
Is it possible to delegate handling of X509 certificate critical extension to custom provider?

A provider for new signature scheme and key type is being developed. When creating X509 certificate, a new critical extension will be included into the certificate (extension comes from openssl conf using `ASN1:SEQUENCE ...`). A correct certificate with a new type of key, signature and extension is created. During verification it should be possible to access data from extension and it should be included into `EVP_PKEY->keydata`. A correct way to achieve that is not found.

Currently there are 2 workarounds:
1) When verifying using homemade code it is possible to implement `verify_cb` to handle unknown critical extension and update `EVP_PKEY->keydata`. It is considered as a hack as `EVP_PKEY` user is not supposed to access keydata directly. Another drawback is that it does not work with OpenSSL tools.

2) Data from those extensions can be included to provider configuration and critical extension can be ignored. It is correct use case but it makes new extension totally pointless and usage of provider will be more complicated due to extra effort to keep provider conf updated.

**Comment by t8m** [comment]
No, we do not yet have a provider interface for X509 related stuff.

**Comment by t8m** [comment]
Do you want to make it an RFE? It would probably require some concrete design proposal how this should be implemented.

**Comment by richsalz** [comment]
Look at X509V#_EXT_add It's unfortunately not documented, but there are sample uses (apps/x509.c, etc) that might be useful. In short, I think what you need is already provided.

**Comment by t8m** [comment]
As the RFE is in #16293 I am closing this one.

**Comment by richsalz** [comment]
It can be done now by adding an exdata to point to the provider, that's my point.

**Comment by beldmit** [comment]
I think it's another beast. We need pluggable X509 extension processing interface, it can be implemented via providers and probably should, but it requires changes in the main libcrypto.
